linenoise/linenoise/linenoise.h

```diff
@@ -0,0 +1,335 @@
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#ifndef __LINENOISE_H
+#define __LINENOISE_H
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#include <sys/types.h>
+#include <unistd.h>
+#include <stdbool.h>
+#include <stddef.h>
+#include <stdint.h>
+#include "esp_err.h"
+
+/*
+ * IMPORTANT: This library is not thread safe.
+ */
+
+/*------------------------------------------------------------------------------------------------
+ *NEW LINENOISE API
+ *------------------------------------------------------------------------------------------------*/
+
+/**
+ * @brief Stores a list of completion strings.
+ */
+typedef struct linenoise_completions {
+  size_t len;   /**< Number of completions. */
+  char **cvec;  /**< Array of completion strings. */
+} linenoise_completions_t;
+
+/**
+ * @brief Callback function for providing completions.
+ *
+ * @param str Input string from the user.
+ * @param lc Completions object to be populated.
+ */
+typedef void (*linenoise_completion_callback)(const char *str, linenoise_completions_t *lc);
+
+/**
+ * @brief Callback function for providing hints during input.
+ *
+ * @param str Input string.
+ * @param color Pointer to set the color of the hint text.
+ * @param bold Pointer to set the bold attribute (non-zero for bold).
+ * @return A dynamically allocated hint string or NULL if no hint is available.
+ */
+typedef char* (*linenoise_hints_callback)(const char *str, int *color, int *bold);
+
+/**
+ * @brief Callback function to free memory returned by hints callback.
+ *
+ * @param ptr Pointer to the hint string to be freed.
+ */
+typedef void (*linenoise_free_hints_callback)(void *ptr);
+
+/**
+ * @brief Function pointer type for reading bytes.
+ *
+ * @param fd File descriptor.
+ * @param buf Buffer to store the read bytes.
+ * @param count Number of bytes to read.
+ * @return Number of bytes read, or -1 on error.
+ */
+typedef ssize_t (*linenoise_read_bytes_fn)(int fd, void* buf, size_t count);
+
+/**
+ * @brief Function pointer type for writing bytes.
+ *
+ * @param fd File descriptor.
+ * @param buf Buffer containing bytes to write.
+ * @param count Number of bytes to write.
+ * @return Number of bytes written, or -1 on error.
+ */
+typedef ssize_t (*linenoise_write_bytes_fn)(int fd, const void* buf, size_t count);
+
+/**
+ * @brief Structure defining the parameters needed by a linenoise
+ * instance
+ */
+typedef struct linenoise_instance_param {
```

Comment:
Instance config structure should have a `void* user_ctx` member that will be passed to all the callbacks.

This design in common in libraries which support multiple instances, and allows, for example, for the caller to find the local "state" or "context" structure related to the instance which has triggered the particular callback. This is also useful when wrapping the library in a C++ class, the `user_ctx` is then often the pointer to `this` (class instance).

Reply:
Implemented.